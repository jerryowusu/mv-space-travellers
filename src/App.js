import {
  HashRouter as Router,
} from 'react-router-dom';
import './App.css';
import Navigator from './components/Navigator';

const App = () => (
  <div className="App">
    <Router>
      <Navigator />
    </Router>
  </div>
);

export default App;
