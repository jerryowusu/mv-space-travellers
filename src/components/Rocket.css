.rocket {
  display: grid;
  grid-auto-flow: column;
  grid-template-columns: 20% 80%;
  gap: 1rem;
  padding: 0;
  box-sizing: border-box;
}

.rocket-image {
  width: 100%;
}

.rocket-name {
  margin: 0;
}

.rocket-info {
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.rocket-button {
  padding: 0.5rem 1rem;
  background-color: #4949f9;
  border: 0;
  border-radius: 4px;
  color: white;
  width: fit-content;
}
