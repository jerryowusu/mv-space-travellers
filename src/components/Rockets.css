.rockets {
  list-style-type: none;
  display: flex;
  flex-direction: column;
  gap: 2rem;
  width: 100%;
  box-sizing: border-box;
  align-items: center;
  justify-content: center;
  padding: 0;
}

.rocket-item {
  width: 90%;
  box-sizing: border-box;
}
