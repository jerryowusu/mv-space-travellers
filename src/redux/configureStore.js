import { combineReducers, createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import rocketsReducer from './rockets/rockets';

const rootReducer = combineReducers({
  rockets: rocketsReducer,
});

const initialState = {};
const middleware = [thunk];
const store = createStore(
  rootReducer,
  initialState,
  composeWithDevTools(applyMiddleware(...middleware)),
);

export default store;
