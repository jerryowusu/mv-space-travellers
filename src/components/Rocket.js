import React from 'react';
import './Rocket.css';

const Rocket = (rocket) => {
  const { name, image, description } = rocket;
  return (
    <div className="rocket">
      <img src={image} alt={name} className="rocket-image" />
      <div className="rocket-info">
        <h2 className="rocket-name">{name}</h2>
        <p className="rocket-description">{description}</p>
        <button type="button" className="rocket-button">Reserve Rocket</button>
      </div>
    </div>
  );
};

export default Rocket;
