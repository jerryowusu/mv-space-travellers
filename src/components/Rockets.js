import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { getRockets } from '../redux/rockets/rockets';
import Rocket from './Rocket';
import './Rockets.css';

const Rockets = (props) => {
  const [rockets, setRockets] = useState([]);
  const rocketsInit = useSelector((state) => state.rockets);
  const dispatch = useDispatch();

  useEffect(() => {
    const { rocketsFetched, setRocketsFetched } = props;
    if (!rocketsFetched) {
      dispatch(getRockets());
      setRocketsFetched(true);
    }
  }, [dispatch, props]);

  useEffect(() => {
    const { rocketsFetched } = props;
    if (rocketsFetched) {
      setRockets(rocketsInit);
    }
  }, [rocketsInit, props]);

  return (
    <ul className="rockets">
      {
        rockets.map((rocket) => (
          <li key={rocket.id} className="rocket-item">
            <Rocket name={rocket.name} image={rocket.image} description={rocket.description} />
          </li>
        ))
      }
    </ul>
  );
};

Rockets.propTypes = {
  rocketsFetched: PropTypes.bool.isRequired,
  setRocketsFetched: PropTypes.func.isRequired,
};

export default Rockets;
