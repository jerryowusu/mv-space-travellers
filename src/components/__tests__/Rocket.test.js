import { screen, render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';
import { combineReducers, createStore, applyMiddleware } from 'redux';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import thunk from 'redux-thunk';
import rocketsReducer from '../../redux/rockets/rockets';
import missionsReducer from '../../redux/missions/missions';
import Rocket from '../Rocket';
import Rockets from '../Rockets';
import App from '../../App';

const mockRocketsData = [
  {
    id: "rocket-1",
    image: "rocket-image-1",
    name: "Rocket 1",
    description: "rocket-description-1",
    reserved: false,
    wikipedia: "rocket-link-1",
  },
  {
    id: "rocket-2",
    image: "rocket-image-2",
    name: "Rocket 2",
    description: "rocket-description-2",
    reserved: false,
    wikipedia: "rocket-link-2",
  },
  {
    id: "rocket-3",
    image: "rocket-image-3",
    name: "Rocket 3",
    description: "rocket-description-3",
    reserved: false,
    wikipedia: "rocket-link-3",
  },
];

const rootReducer = combineReducers({
  rockets: rocketsReducer,
  missions: missionsReducer,
});

const store = createStore(
  rootReducer,
  applyMiddleware(thunk),
);

const mockGetRockets = () => ({
  type: 'GET_ROCKETS',
  rockets: mockRocketsData,
});

store.dispatch(mockGetRockets());

describe('Rocket component tests:', () => {
  it('Component renders correctly', () => {
    const reserveRocketAction = jest.fn();
    const cancelReservationAction = jest.fn();
    const rocket = renderer.create(
      <Rocket
        id={mockRocketsData[0].id}
        name={mockRocketsData[0].name}
        image={mockRocketsData[0].image}
        description={mockRocketsData[0].description}
        reserveRocket={reserveRocketAction}
        cancelReservation={cancelReservationAction}
        reserved={mockRocketsData[0].reserved}
      />,
    ).toJSON();
    expect(rocket).toMatchSnapshot();
  });

  it('Display name of the rocket correctly', () => {
    const reserveRocketAction = jest.fn();
    const cancelReservationAction = jest.fn();
    const { container } = render(
      <Rocket
        id={mockRocketsData[0].id}
        name={mockRocketsData[0].name}
        image={mockRocketsData[0].image}
        description={mockRocketsData[0].description}
        reserveRocket={reserveRocketAction}
        cancelReservation={cancelReservationAction}
        reserved={mockRocketsData[0].reserved}
      />
    );
    const name = container.querySelector('.rocket-name');
    expect(name.textContent).toEqual(mockRocketsData[0].name);
  });

  it('Display image of the rocket correctly', () => {
    const reserveRocketAction = jest.fn();
    const cancelReservationAction = jest.fn();
    const { container } = render(
      <Rocket
        id={mockRocketsData[0].id}
        name={mockRocketsData[0].name}
        image={mockRocketsData[0].image}
        description={mockRocketsData[0].description}
        reserveRocket={reserveRocketAction}
        cancelReservation={cancelReservationAction}
        reserved={mockRocketsData[0].reserved}
      />
    );
    const image = container.querySelector('.rocket-image');
    expect(image.src).toEqual(`http://localhost/${mockRocketsData[0].image}`);
  });

  it('Reserve the first rocket correctly (check if badge exist)', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    fireEvent.click(screen.getAllByText('Reserve Rocket')[0]);
    expect(screen.getByText('Reserved')).toBeInTheDocument();
  });

  it('Reserve the first rocket correctly (check if button changed)', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    fireEvent.click(screen.getAllByText('Cancel Reservation')[0]);
    fireEvent.click(screen.getAllByText('Reserve Rocket')[0]);
    expect(screen.getByText('Cancel Reservation')).toBeInTheDocument();
  });

  it('Cancel reservation for the first rocket correctly', () => {
    render(
      <Provider store={store}>
        <App />
      </Provider>
    );
    fireEvent.click(screen.getAllByText('Cancel Reservation')[0]);
    expect(store.getState().rockets[0].reserved).toEqual(false);
  });
});