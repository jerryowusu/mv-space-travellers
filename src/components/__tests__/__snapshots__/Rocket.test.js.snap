// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rocket component tests: Component renders correctly 1`] = `
<div
  className="rocket"
>
  <img
    alt="Rocket 1"
    className="rocket-image"
    src="rocket-image-1"
  />
  <div
    className="rocket-info"
  >
    <h2
      className="rocket-name"
    >
      Rocket 1
    </h2>
    <div
      className="rocket-description-container"
    >
      <p
        className="rocket-description"
      >
        rocket-description-1
      </p>
    </div>
    <button
      className="rocket-button unclicked"
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;
